src/llama-kv-cells.h

```diff
@@ -0,0 +1,259 @@
+    bool seq_keep(uint32_t i, llama_seq_id seq_id) {
+        assert(i < pos.size());
+
+        if (seq[i].test(seq_id)) {
+            seq[i].reset();
+            seq[i].set(seq_id);
+
+            return false;
+        }
+
+        if (seq[i].any()) {
+            seq[i].reset();
+            pos[i] = -1;
+
+            used--;
+
+            return true;
+        }
+
+        assert(pos[i] == -1);
+
+        return false;
+    }
+
+    bool seq_has(uint32_t i, llama_seq_id seq_id) const {
+        assert(i < pos.size());
+        assert(seq_id >= 0);
+
+        return seq[i].test(seq_id);
+    }
+
+    // note: call only if the cell is not empty
+    bool seq_add(uint32_t i, llama_seq_id seq_id) {
+        assert(i < pos.size());
+        assert(pos[i] != -1);
+
+        if (seq[i].none()) {
+            seq[i].set(seq_id);
+
+            used++;
+
+            return true;
+        }
+
+        return false;
+    }
+
+    // note: call only if the cell is not empty
+    llama_pos get_pos(uint32_t i) const {
+        assert(i < pos.size());
+        assert(pos[i] != -1);
+
+        return pos[i];
+    }
+
+    // note: call only if the cell is not empty
+    llama_pos get_delta(uint32_t i) const {
+        assert(i < pos.size());
+        assert(pos[i] != -1);
+
+        return delta[i];
+    }
+
+    bool pos_in(uint32_t i, llama_pos p0, llama_pos p1) const {
+        assert(i < pos.size());
+
+        return pos[i] >= p0 && pos[i] < p1;
+    }
+
+    // note: call only if the cell is empty
+    void pos_set(uint32_t i, llama_pos p) {
+        assert(i < pos.size());
+        assert(pos[i] == -1);
+
+        pos[i] = p;
+        used++;
+    }
+
+    // pos[i] = pos[i] + d
+    // note: call only if the cell is not empty
+    bool pos_add(uint32_t i, llama_pos d) {
+        assert(i < pos.size());
+        assert(pos[i] != -1);
+
+        pos[i]   += d;
+        delta[i] += d;
+
+        has_delta = true;
+
+        if (pos[i] < 0) {
+            pos[i] = -1;
+            seq[i].reset();
+
+            used--;
+
+            return true;
+        }
+
+        return false;
+    }
+
+    // pos[i] = pos[i] / d
+    // note: call only if the cell is not empty
+    void pos_div(uint32_t i, int d) {
+        assert(i < pos.size());
+        assert(pos[i] != -1);
+
+        const llama_pos p_old = pos[i];
+
+        pos[i]   /= d;
+        delta[i] += p_old - pos[i];
+
+        has_delta = true;
+    }
+
+    bool pos_has_shift() const {
+        return has_delta;
+    }
+
+    void pos_reset_delta() {
+        has_delta = false;
+
+        for (uint32_t i = 0; i < delta.size(); ++i) {
+            delta[i] = 0;
+        }
+    }
+
+private:
+    uint32_t used = 0; // used cells (i.e. at least one seq_id)
+
+    bool has_delta = false;
+
+    std::vector<llama_pos> pos;
+    std::vector<llama_pos> delta;
+
+    // TODO: assert n_seq_max <= 64
+    std::vector<std::bitset<64>> seq;
```

Comment:
If there is going to be a limit to `n_seq_max` now, it should be documented and the check should be added now, ideally at the time the context is created. Otherwise this will eventually lead to hard to diagnose bugs.

Reply:
Introduced `llama_max_parallel_sequences()` in [0a8cdc3](https://github.com/ggml-org/llama.cpp/pull/13706/commits/0a8cdc3ad4082976e7dad86a988cee4c2547a279). If the user attempts to create a context with a larger number, it will be clamped to the max allowed.

Reply:
I don't think clamping the value here is enough, the application needs to know how many sequence ids it can use. It should return an error.

Reply:
Changed to an error in [b495877](https://github.com/ggml-org/llama.cpp/pull/13706/commits/b4958772bdb19f10d2dc5af2c27156e9662e7a48)